crates/core/src/vm.rs

```diff
@@ -328,26 +354,32 @@ impl<'db, 'tx> DbProgram<'db, 'tx, MutTx> {
     }
 
     fn _execute_insert(&mut self, table: &Table, rows: Vec<ProductValue>) -> Result<Code, ErrorVm> {
-        match table {
-            // TODO: How do we deal with mutating values?
-            Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
-            Table::DbTable(x) => {
-                for row in rows {
-                    self.db.insert(self.tx, x.table_id, row)?;
+        match self.tx {
+            TxMode::MutTx(tx) => match table {
+                // TODO: How do we deal with mutating values?
+                Table::MemTable(_) => Err(ErrorVm::Other(anyhow::anyhow!("How deal with mutating values?"))),
+                Table::DbTable(x) => {
+                    for row in rows {
+                        self.db.insert(tx, x.table_id, row)?;
+                    }
+                    Ok(Code::Pass)
                 }
-                Ok(Code::Pass)
-            }
+            },
+            TxMode::Tx(_) => unreachable!("mutable operation is invalid with read tx"),
```

Comment:
If here is this then the upper layer should prove this is not possible. If can't it should be a normal `Err` instead (is better to avoid a crash)

Reply:
Reaching to this branch anytime implies a bug in code and not to do anything with Users interaction.
I added a comment on `VmProgram::eval_query()` (only caller) to specify safetu gurantee.